File: android/diag/src/main/kotlin/com/fieldtap/diag/Hdlc.kt
package com.fieldtap.diag

/**
 * The framing Qualcomm's diag protocol uses on the wire and in a `.qmdl` file.
 *
 * A frame is its payload, then a 16-bit CRC, then `0x7E`. Inside, `0x7D` and `0x7E` are escaped as
 * `0x7D 0x5D` and `0x7D 0x5E`. The CRC is CRC-16/X-25: reflected, initial value `0xFFFF`, final XOR
 * `0xFFFF`, so a frame including its own CRC checks to the constant [GOOD_CRC].
 *
 * This is a port of `fieldtap/diag/hdlc.py` and must stay byte-for-byte compatible with it: the
 * handset writes files the desktop tool reads, and the same capture has to decode identically on
 * both. [crc16] is checked against the published CRC-16/X-25 check value in the tests, which is the
 * one number that pins the whole table.
 *
 * Owner: workstream `diag-on-handset`.
 */
object Hdlc {

    const val FLAG: Byte = 0x7E
    const val ESCAPE: Byte = 0x7D
    private const val ESCAPE_MASK = 0x20

    /** What [crc16] returns for a buffer that already ends in its own correct CRC. */
    const val GOOD_CRC = 0x0F47

    private val TABLE = IntArray(256).also { table ->
        for (byte in 0 until 256) {
            var crc = byte
            repeat(8) { crc = if (crc and 1 != 0) (crc ushr 1) xor 0x8408 else crc ushr 1 }
            table[byte] = crc
        }
    }

    /** CRC-16/X-25 over [data]. */
    fun crc16(data: ByteArray, from: Int = 0, until: Int = data.size): Int {
        var crc = 0xFFFF
        for (i in from until until) {
            crc = (crc ushr 8) xor TABLE[(crc xor (data[i].toInt() and 0xFF)) and 0xFF]
        }
        return crc.inv() and 0xFFFF
    }

    /** [payload] as a complete frame: payload, CRC, trailing flag, with escaping applied. */
    fun encode(payload: ByteArray): ByteArray {
        val crc = crc16(payload)
        val body = payload + byteArrayOf((crc and 0xFF).toByte(), ((crc ushr 8) and 0xFF).toByte())
        val out = ArrayList<Byte>(body.size + 8)
        for (b in body) {
            if (b == FLAG || b == ESCAPE) {
                out.add(ESCAPE)
                out.add((b.toInt() xor ESCAPE_MASK).toByte())
            } else {
                out.add(b)
            }
        }
        out.add(FLAG)
        return ByteArray(out.size) { out[it] }
    }

    /** [data] with escape sequences resolved. A trailing lone escape is dropped. */
    fun unescape(data: ByteArray, from: Int = 0, until: Int = data.size): ByteArray {
        val out = ByteArray(until - from)
        var n = 0
        var i = from
        while (i < until) {
            val b = data[i]
            if (b == ESCAPE) {
                if (i + 1 >= until) break
                out[n++] = (data[i + 1].toInt() xor ESCAPE_MASK).toByte()
                i += 2
            } else {
                out[n++] = b
                i++
            }
        }
        return out.copyOf(n)
    }
}

/** A frame that did not survive its own CRC. */
class HdlcError(message: String) : IllegalArgumentException(message)

/**
 * Splits a byte stream into frames, keeping whatever is incomplete until more arrives.
 *
 * [feed] may be called with any split of the stream — one byte at a time or a whole file — and
 * returns the same frames either way. Frames whose CRC fails are counted in [crcErrors] and dropped
 * rather than thrown, because one corrupt frame must not end a capture.
 */
class Unframer {
    private val buffer = ArrayList<Byte>(4096)

    /** Frames dropped because their CRC did not check. */
    var crcErrors: Int = 0
        private set

    /** Bytes held for a frame that has not ended yet. */
    val pending: Int get() = buffer.size

    fun feed(data: ByteArray): List<ByteArray> {
        val frames = ArrayList<ByteArray>()
        for (b in data) {
            if (b == Hdlc.FLAG) {
                if (buffer.isNotEmpty()) {
                    decodeOrNull(ByteArray(buffer.size) { buffer[it] })?.let { frames.add(it) }
                    buffer.clear()
                }
            } else {
                buffer.add(b)
            }
        }
        return frames
    }

    /** The payload of one escaped frame body, or null when its CRC does not check. */
    private fun decodeOrNull(escaped: ByteArray): ByteArray? {
        val raw = Hdlc.unescape(escaped)
        if (raw.size < 3) {
            crcErrors++
            return null
        }
        if (Hdlc.crc16(raw) != Hdlc.GOOD_CRC) {
            crcErrors++
            return null
        }
        return raw.copyOf(raw.size - 2)
    }
}

File: ios/FieldTapKit/Sources/FTCore/Hdlc.swift
// Port of android/diag/src/main/kotlin/com/fieldtap/diag/Hdlc.kt.

import Foundation

/// The framing Qualcomm's DIAG protocol uses on the wire and in a `.qmdl` file.
///
/// A frame is its payload, then a 16-bit CRC, then 0x7E. Inside, 0x7D and 0x7E are escaped as 0x7D 0x5D and
/// 0x7D 0x5E. The CRC is CRC-16/X-25 (reflected, init 0xFFFF, final XOR 0xFFFF), so a frame including its own
/// CRC checks to `goodCrc`. Must stay byte-compatible with fieldtap/diag/hdlc.py and the Kotlin port: the
/// same capture has to decode identically everywhere.
public enum Hdlc {
    public static let flag: UInt8 = 0x7E
    public static let escape: UInt8 = 0x7D
    static let escapeMask: UInt8 = 0x20

    /// What `crc16` returns for a buffer that already ends in its own correct CRC.
    public static let goodCrc = 0x0F47

    static let table: [UInt16] = (0..<256).map { byte in
        var crc = UInt16(byte)
        for _ in 0..<8 { crc = (crc & 1) != 0 ? (crc >> 1) ^ 0x8408 : crc >> 1 }
        return crc
    }

    /// CRC-16/X-25 over `data`.
    public static func crc16<C: Collection>(_ data: C) -> Int where C.Element == UInt8 {
        if let crc = data.withContiguousStorageIfAvailable({ crc16($0) }) { return crc }
        return Array(data).withUnsafeBufferPointer { crc16($0) }
    }

    /// The non-generic loop every overload ends in (a Debug build does not specialise generics).
    public static func crc16(_ p: UnsafeBufferPointer<UInt8>) -> Int {
        var crc: UInt16 = 0xFFFF
        table.withUnsafeBufferPointer { t in
            for i in 0..<p.count { crc = (crc >> 8) ^ t[Int((crc ^ UInt16(p[i])) & 0xFF)] }
        }
        return Int(~crc)
    }

    /// `payload` as a complete frame: payload, CRC, trailing flag, with escaping applied.
    public static func encode(_ payload: [UInt8]) -> [UInt8] {
        let crc = crc16(payload)
        var out: [UInt8] = []
        out.reserveCapacity(payload.count + 8)
        for b in payload + [UInt8(crc & 0xFF), UInt8((crc >> 8) & 0xFF)] {
            if b == flag || b == escape {
                out.append(escape)
                out.append(b ^ escapeMask)
            } else {
                out.append(b)
            }
        }
        out.append(flag)
        return out
    }

    /// `data` with escape sequences resolved. A trailing lone escape is dropped.
    public static func unescape<C: Collection>(_ data: C) -> [UInt8] where C.Element == UInt8 {
        if let out = data.withContiguousStorageIfAvailable({ unescape($0) }) { return out }
        return Array(data).withUnsafeBufferPointer { unescape($0) }
    }

    public static func unescape(_ p: UnsafeBufferPointer<UInt8>) -> [UInt8] {
        [UInt8](unsafeUninitializedCapacity: p.count) { out, n in
            var i = 0
            n = 0
            while i < p.count {
                let b = p[i]
                if b == escape {
                    if i + 1 >= p.count { break }
                    out[n] = p[i + 1] ^ escapeMask
                    i += 2
                } else {
                    out[n] = b
                    i += 1
                }
                n += 1
            }
        }
    }
}

/// Splits a byte stream into frames, keeping whatever is incomplete until more arrives.
///
/// `feed` may be called with any split of the stream, one byte at a time or a whole file, and returns the
/// same frames either way. Frames whose CRC fails are counted in `crcErrors` and dropped rather than thrown,
/// because one corrupt frame must not end a capture. Contiguous input is scanned for flags with memchr, so
/// a 40 MB capture unframes in well under a second even in a Debug build.
public struct Unframer: Sendable {
    private var buffer: [UInt8] = []

    /// Frames dropped because their CRC did not check.
    public private(set) var crcErrors = 0

    /// Bytes held for a frame that has not ended yet.
    public var pending: Int { buffer.count }

    public init() { buffer.reserveCapacity(4096) }

    public mutating func feed<C: Collection>(_ data: C) -> [[UInt8]] where C.Element == UInt8 {
        if let frames = data.withContiguousStorageIfAvailable({ feedBuffer($0) }) { return frames }
        return Array(data).withUnsafeBufferPointer { feedBuffer($0) }
    }

    private mutating func feedBuffer(_ p: UnsafeBufferPointer<UInt8>) -> [[UInt8]] {
        guard let base = p.baseAddress, !p.isEmpty else { return [] }
        var frames: [[UInt8]] = []
        var start = 0
        while start < p.count {
            guard let hit = memchr(base + start, Int32(Hdlc.flag), p.count - start) else {
                buffer.append(contentsOf: UnsafeBufferPointer(rebasing: p[start...]))
                break
            }
            let flagAt = base.distance(to: hit.assumingMemoryBound(to: UInt8.self))
            if buffer.isEmpty {
                if flagAt > start, let f = decodeOrNil(UnsafeBufferPointer(rebasing: p[start..<flagAt])) {
                    frames.append(f)
                }
            } else {
                buffer.append(contentsOf: UnsafeBufferPointer(rebasing: p[start..<flagAt]))
                let held = buffer
                buffer.removeAll(keepingCapacity: true)
                if let f = held.withUnsafeBufferPointer({ decodeOrNil($0) }) { frames.append(f) }
            }
            start = flagAt + 1
        }
        return frames
    }

    /// The payload of one escaped frame body, or nil when its CRC does not check.
    private mutating func decodeOrNil(_ escaped: UnsafeBufferPointer<UInt8>) -> [UInt8]? {
        var raw = Hdlc.unescape(escaped)
        guard raw.count >= 3, Hdlc.crc16(raw) == Hdlc.goodCrc else {
            crcErrors += 1
            return nil
        }
        raw.removeLast(2)
        return raw
    }
}
